psh/wget/wget.c

```diff
@@ -0,0 +1,500 @@
+	}
+	printf("Connected\n");
+
+	freeaddrinfo(res);
+
+	return fd;
+}
+
+
+static int wget_request(void)
+{
+	char buf[256];
+	int bytes = 0, i = 0, ret;
+
+	bytes += snprintf(buf, sizeof(buf), "GET /%s HTTP/1.1\r\n", common.path);
+	if (bytes >= sizeof(buf)) {
+		fprintf(stderr, "wget: Request url too large!\n");
+		return -1;
+	}
+
+	bytes += snprintf(buf + bytes, sizeof(buf) - bytes, "Host: %s\r\nUser-Agent: Wget\r\n", common.host);
+	if (bytes >= sizeof(buf)) {
+		fprintf(stderr, "wget: Request url too large!\n");
+		return -1;
+	}
+
+	bytes += snprintf(buf + bytes, sizeof(buf) - bytes, "\r\n");
+	if (bytes > sizeof(buf)) {
+		fprintf(stderr, "wget: Request url too large!\n");
+		return -1;
+	}
+
+	while (i < bytes) {
+		ret = write(common.socket, buf + i, bytes - i);
+		if (ret < 0) {
+			return -1;
+		}
+
+		i += ret;
+	}
+
+	if (common.debug) {
+		printf("--------Request--------\n%*s\n", bytes, buf);
+	}
+
+	return 0;
+}
+
+
+static int wget_file(size_t len)
+{
+	int ret;
+	size_t left = len;
+	int i = 0;
+
+	/* Save what is left in the buffer */
+	if (common.readptr < common.buf + common.len) {
+		ret = fwrite(common.readptr, 1, common.buf + common.len - common.readptr, common.outfile);
+		if (ret < 0) {
+			fprintf(stderr, "wget: Write failed\n");
+			return -1;
+		}
+		left -= ret;
+	}
+
+	while (left > 0) {
+		ret = read(common.socket, common.buf, min(sizeof(common.buf), left));
+		if (ret < 0) {
+			fprintf(stderr, "wget: read from socket failed!\n");
+			return -1;
+		}
+
+		left -= ret;
+
+		ret = fwrite(common.buf, 1, ret, common.outfile);
+		if (ret < 0) {
+			fprintf(stderr, "wget: Write failed\n");
+			return -1;
+		}
+		if (i == 0) {
+			printf("\rWritten: %8d/%d", len - left, len);
+			fflush(stdout);
+		}
+		i = (i + 1) % 8;
+	}
+	printf("\rWritten: %8d/%d", len - left, len);
+	fflush(stdout);
+	putchar('\n');
+
+	return len - left;
+}
+
+
+static int wget_buf(void)
+{
+	int ret;
+
+	/* Move not consumed part of the buffer */
+	common.len = common.len - (common.readptr - common.buf);
+	memmove(common.buf, common.readptr, common.len);
+	common.readptr = common.buf;
+
+	ret = read(common.socket, common.buf + common.len, sizeof(common.buf) - common.len);
+	if (ret <= 0) {
+		return -1;
+	}
+
+	common.len += ret;
+
+	return 0;
+}
+
+
+static char *wget_hdrnext(void)
+{
+	char *hdr = NULL, *tmp;
+
+	while (hdr == NULL) {
+		tmp = memchr(common.readptr, '\n', common.buf + common.len - common.readptr);
+		if (tmp == NULL) {
+			/* Carry on reading */
+			if (wget_buf() != 0) {
+				return NULL;
+			}
+			continue;
+		}
+
+		*(tmp - 1) = '\0';
+		hdr = common.readptr;
+		common.readptr = tmp + 1;
+	}
+
+	return hdr;
+}
+
+
+static size_t wget_parsehdrs(void)
+{
+	char *hdr = NULL, *endptr, *num;
+	size_t contentLen = 0;
+
+	/* Read headers until no more data is available or empty line is found */
+	do {
+		hdr = wget_hdrnext();
+		if (hdr == NULL) {
+			return -1;
+		}
+
+		if (strncasecmp(hdr, "content-length: ", 15) == 0) {
```

Comment:
```suggestion
		if (strncasecmp(hdr, "content-length: ", 16) == 0) {
```
or 15 with no space at the end of the string.